src/metabase/lib/portable.cljc

```diff
@@ -0,0 +1,77 @@
+(ns metabase.lib.portable
+  (:require
+   [medley.core :as m]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.lib.walk :as lib.walk]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]))
+
+(defprotocol QueryVisitor
+  (visit-database-id [this mp context database-id])
+  (visit-table-id    [this mp context source-table-id])
+  (visit-card-id     [this mp context source-card-id])
+  (visit-field-id    [this mp context field-id]))
+
+(defn query-visitor?
+  "Does x implement the QueryVisitor protocol?"
+  [x]
+  #?(:clj  (extends? QueryVisitor (class x))
+     :cljs (satisfies? QueryVisitor x)))
+
+(mr/def ::query-visitor
+  "Schema for things that satisfy the [[metabase.lib.portable/QueryVisitor]] protocol."
+  [:fn
+   {:error/message "Valid QueryVisitor"}
+   #'query-visitor?])
+
+(defn- update-field [mp k f]
+  (m/assoc-some mp k (some-> (get mp k) f)))
+
+(defn- walk-database [query context visitor]
+  (let [mp (lib.metadata/->metadata-provider query)]
+    (update-field query :database (partial visit-database-id visitor mp context))))
+
+(defn- walk-stages [query context visitor]
+  (let [mp (lib.metadata/->metadata-provider query)
+        st (partial visit-table-id visitor mp context)
+        sc (partial visit-card-id  visitor mp context)]
+    (mu/disable-enforcement
+      (lib.walk/walk-stages
+       query
+       (fn [_q _pt stage]
+         (-> stage
+             (update-field :source-table st)
+             (update-field :source-card  sc)))))))
+
+(defn- walk-clauses [query context visitor]
+  (let [mp (lib.metadata/->metadata-provider query)]
+    (mu/disable-enforcement
+      (lib.walk/walk-clauses
+       query
+       (fn [_q _pt _sjp clause]
+         (lib.util.match/match-lite clause
+           [:field opts id]
+           (when-some [field-ref (visit-field-id visitor mp context id)]
```

Comment:
this seems weird, what are the cases where `visit-field-id` doesn't return something? Replacing a `:field` clause with `nil` will probably result in an invalid query